keps/sig-node/4210-max-image-gc-age/README.md

```diff
@@ -0,0 +1,373 @@
+However, there are cases that additional conditions could be considered useful. One such condition is maximum age of an image.
+If an image is unused for a long time (the exact amount of time will be decided, but on the order of weeks is what comes to mind),
+then it is not likely to be used again.
+
+One such condition that can be imagined are clusters with automatic upgrades. If a cluster goes through an upgrade process,
+it will likely have images cached from the old release (old kube-apiserver/etcd/etc). While these images would eventually get removed
+through the disk usage condition, they will needlessly occupy disk space before that.
+
+### Goals
+
+- Introduce an option to the Kubelet ImageMaximumGCAge
+
+### Non-Goals
+
+- Introduce other conditions for image garbage collection to trigger.
+    - A WG was put together in SIG-Node to collect other GC use cases, and no other conditions were identified.
+
+## Proposal
+
+Kubelet has three different configuration fields for image garbage collection:
+- ImageMinimumGCAge: the youngest an image can be to be qualified for garbage collection
+- ImageGCLowThresholdPercent: The lowest disk usage will be before garbage collection begins
+- ImageGCHighThresholdPercent: The highest disk usage will be before garbage collection runs each GC period.
+
+Between each of these options, there is a common thread: image garbage collection only triggers when disk usage has reached a certain threshold.
+In other words, there are no alternative conditions the Kubelet will begin triggering collection. Luckily, this condition is the most important:
+the primary goal of garbage collection is to ensure images don't clutter the disk too much and cause it to fill up needlessly.
+However, having the Kubelet be purely reactive means that images will clutter the disk and cause it to fill up. While there aren't reported cases
+where this causes issues, it is inefficient with disk space and can cause the Kubelet to scramble to save disk space when the threshold is met.
+
+An additional approach is to define a way for an admin to request images are cleaned up after they're unused for a certain period of time.
+This would reduce the frequency of the disk usage hitting the level, and provide an admin more flexibility in how garbage collection is defined.
+
+The proposal of this KEP is to add an option to the KubeletConfiguration object that looks like:
+```
+	// ImageMaximumGCAge is the maximum age an image can be unused before it is garbage collected.
+    // The default of this field is 0, which disables it.
+    // +optional
+	ImageMaximumGCAge metav1.Duration
+```
+
+To begin, this option will be set to 0, which will be interpreted as "disabled". In the future, a more reasonable default may be chosen.
+
+### User Stories (Optional)
+
+#### Story 1
+
+- As a cluster admin, I would like my unused images to be garbage collected in a timely manner, and not occupy disk space forever.
```

Comment:
do we provide guarantee or just a best effort?

Reply:
I don't think we (kubelet) can promise anything as it's ultimately done by the CRI

Reply:
Let's explicitly say that the Story 1 is not about guaranteed clean up, but the best effort.